service/src/main/java/io/camunda/service/exception/ErrorMapper.java

```diff
@@ -0,0 +1,296 @@
+  }
+
+  public static ServiceException mapBrokerRejection(final BrokerRejection rejection) {
+    return new ServiceException(mapRejectionToServiceError(rejection));
+  }
+
+  public static ServiceException mapDocumentError(final DocumentError documentError) {
+    return new ServiceException(mapDocumentErrorToServiceError(documentError));
+  }
+
+  public static ServiceException mapSearchError(final CamundaSearchException cse) {
+    final String errorMessage = cse.getMessage();
+    return switch (cse.getReason()) {
+      case NOT_FOUND -> new ServiceException(errorMessage, NOT_FOUND);
+      case NOT_UNIQUE -> new ServiceException(errorMessage, ALREADY_EXISTS);
+      case CONNECTION_FAILED -> {
+        final String detail = "The search client could not connect to the search server";
+        LOGGER.debug(detail, cse);
+        yield new ServiceException(detail, UNAVAILABLE);
+      }
+      case SEARCH_SERVER_FAILED -> {
+        final String detail = "The search server was unable to process the request";
+        LOGGER.debug(detail, cse);
+        yield new ServiceException(detail, INTERNAL);
+      }
+      case SEARCH_CLIENT_FAILED -> {
+        final String detail = "The search client was unable to process the request";
+        LOGGER.debug(detail, cse);
+        yield new ServiceException(detail, INTERNAL);
+      }
+      default -> {
+        LOGGER.debug(errorMessage, cse);
+        yield new ServiceException(errorMessage, INTERNAL);
+      }
+    };
+  }
+
+  public static ServiceException createForbiddenException(final Authorization authorization) {
+    return new ServiceException(
+        "Unauthorized to perform operation '%s' on resource '%s'"
+            .formatted(authorization.permissionType(), authorization.resourceType()),
+        FORBIDDEN);
+  }
+
+  private static ServiceError mapErrorToServiceError(final Throwable error) {
+    return mapErrorToServiceError(error, error, ErrorMapper.LOGGER);
+  }
+
+  private static ServiceError mapErrorToServiceError(
+      final Throwable rootError, final Throwable error, final Logger logger) {
+    final var builder = ServiceError.newBuilder();
+
+    switch (error) {
+      case final ExecutionException e -> {
+        return mapErrorToServiceError(rootError, e.getCause(), logger);
+      }
+      case final CompletionException e -> {
+        return mapErrorToServiceError(rootError, e.getCause(), logger);
+      }
+      case final BrokerErrorException brokerError -> {
+        final ServiceError serviceError =
+            mapBrokerErrorToServiceError(rootError, brokerError.getError(), logger);
+        builder.mergeFrom(serviceError);
+        logger.trace("Expected to handle request, but the broker rejected it", rootError);
+      }
+      case final BrokerRejectionException rejection -> {
+        final ServiceError serviceError = mapRejectionToServiceError(rejection.getRejection());
+        builder.mergeFrom(serviceError);
```

Comment:
🤔 It feels like we're doing duplicate work since we're using an instance of the `ServiceError.Builder` within the "mapper" methods, and then copying the output into another instance of the same Builder.

Can't we simplify the "mapper" methods by passing the initial `Builder` instance? We wouldn't need the `mergeFrom` method then.

Reply:
I think the initial `builder` is not even needed, so we don't need any merge or copy. I'll see what I can do here to simplify this, good call 👍

Reply:
Done with https://github.com/camunda/camunda/pull/35207/commits/98bc22a716e5cb8708b4088abf56fe57d8292e6b